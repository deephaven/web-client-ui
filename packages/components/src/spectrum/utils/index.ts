export * from './itemUtils';
export * from './itemWrapperUtils';
export * from './propsUtils';
export * from './themeUtils';
export * from './useRenderNormalizedItem';
export * from './useStringifiedMultiSelection';
export * from './useStringifiedSelection';
