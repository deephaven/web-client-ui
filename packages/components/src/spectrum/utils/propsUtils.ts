import type { AriaLabelingProps, StyleProps } from '@react-types/shared';

/**
 * Separate props for Spectrum components into AriaLabelingProps, StyleProps, and
 * any remaining props.
 * @param props The props to separate
 * @returns The separated props
 */
export function separateSpectrumProps<T extends AriaLabelingProps & StyleProps>(
  props: T
): {
  ariaLabelProps: AriaLabelingProps;
  styleProps: StyleProps;
  componentProps: Omit<T, keyof (AriaLabelingProps & StyleProps)>;
} {
  const {
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledby,
    'aria-describedby': ariaDescribedby,
    'aria-details': ariaHidden,

    marginX,
    marginY,
    width,
    height,
    minWidth,
    minHeight,
    maxWidth,
    maxHeight,
    flex,
    flexGrow,
    flexShrink,
    flexBasis,
    justifySelf,
    alignSelf,
    order,
    gridArea,
    gridColumn,
    gridRow,
    gridColumnStart,
    gridColumnEnd,
    gridRowStart,
    gridRowEnd,
    position,
    zIndex,
    top,
    bottom,
    start,
    end,
    left,
    right,
    isHidden,
    ...componentProps
  } = props;

  return {
    ariaLabelProps: {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledby,
      'aria-describedby': ariaDescribedby,
      'aria-details': ariaHidden,
    },
    styleProps: {
      marginX,
      marginY,
      width,
      height,
      minWidth,
      minHeight,
      maxWidth,
      maxHeight,
      flex,
      flexGrow,
      flexShrink,
      flexBasis,
      justifySelf,
      alignSelf,
      order,
      gridArea,
      gridColumn,
      gridRow,
      gridColumnStart,
      gridColumnEnd,
      gridRowStart,
      gridRowEnd,
      position,
      zIndex,
      top,
      bottom,
      start,
      end,
      left,
      right,
      isHidden,
    },
    componentProps: componentProps as Omit<
      T,
      keyof (AriaLabelingProps & StyleProps)
    >,
  };
}

export default separateSpectrumProps;
