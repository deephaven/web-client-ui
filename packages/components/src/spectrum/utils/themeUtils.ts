import { theme } from '@react-spectrum/theme-default';
import { themeSpectrumClassesCommon } from '../../theme/theme-spectrum';

export {
  useStyleProps,
  baseStyleProps,
  viewStyleProps,
} from '@react-spectrum/utils';

const { global, light, dark, medium, large } = theme;

/**
 * Extend light + dark theme variables with DH defaults.
 *
 * A theme is just a mapped collection of css class names that are generated
 * from a collection of css modules.
 *
 * e.g.
 * {
 *   global: {
 *     spectrum: 'spectrum_9e130c',
 *     'spectrum--medium': 'spectrum--medium_9e130c',
 *     'spectrum--large': 'spectrum--large_9e130c',
 *     'spectrum--darkest': 'spectrum--darkest_9e130c',
 *     'spectrum--dark': 'spectrum--dark_9e130c',
 *     'spectrum--light': 'spectrum--light_9e130c',
 *     'spectrum--lightest': 'spectrum--lightest_9e130c',
 *   },
 *   light: {
 *     'spectrum--light': 'spectrum--light_a40724',
 *     'dh-spectrum-theme--light': '_dh-spectrum-theme--light_1hblg_22',
 *   },
 *   dark: {
 *     'spectrum--darkest': 'spectrum--darkest_256eeb',
 *     'dh-spectrum-theme--dark': '_dh-spectrum-theme--dark_f7kge_22',
 *   },
 *   medium: {
 *     'spectrum--medium': 'spectrum--medium_4b172c',
 *   },
 *   large: {
 *     'spectrum--large': 'spectrum--large_c40598',
 *   },
 * }
 */
/* eslint-disable import/prefer-default-export */
export const themeDHDefault = {
  global,
  light: {
    ...light,
    ...themeSpectrumClassesCommon,
  },
  dark: {
    ...dark,
    ...themeSpectrumClassesCommon,
  },
  // scales
  medium,
  large,
};
