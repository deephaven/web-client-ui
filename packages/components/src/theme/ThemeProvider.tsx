import { createContext, ReactNode, useEffect, useMemo, useState } from 'react';
import Log from '@deephaven/log';
import { DEFAULT_DARK_THEME_KEY, ThemeData } from './ThemeModel';
import {
  calculatePreloadStyleContent,
  getActiveThemes,
  getDefaultBaseThemes,
  getThemePreloadData,
  setThemePreloadData,
  overrideSVGFillColors,
} from './ThemeUtils';
import { SpectrumThemeProvider } from './SpectrumThemeProvider';
import './theme-svg.scss';

export interface ThemeContextValue {
  activeThemes: ThemeData[] | null;
  selectedThemeKey: string;
  themes: ThemeData[];
  setSelectedThemeKey: (themeKey: string) => void;
}

const log = Log.module('ThemeProvider');

export const ThemeContext = createContext<ThemeContextValue | null>(null);

export interface ThemeProviderProps {
  /*
   * Additional themes to load in addition to the base themes. If no additional
   * themes are to be loaded, this must be set to an empty array in order to
   * tell the provider to activate the base themes.
   */
  themes: ThemeData[] | null;
  children: ReactNode;
}

export function ThemeProvider({
  themes: customThemes,
  children,
}: ThemeProviderProps): JSX.Element | null {
  const baseThemes = useMemo(() => getDefaultBaseThemes(), []);

  const [value, setValue] = useState<ThemeContextValue | null>(null);

  const [selectedThemeKey, setSelectedThemeKey] = useState<string>(
    () => getThemePreloadData()?.themeKey ?? DEFAULT_DARK_THEME_KEY
  );

  // Calculate active themes once a non-null themes array is provided.
  const activeThemes = useMemo(
    () =>
      customThemes == null
        ? null
        : getActiveThemes(selectedThemeKey, {
            base: baseThemes,
            custom: customThemes ?? [],
          }),
    [baseThemes, selectedThemeKey, customThemes]
  );

  const themes = useMemo(
    () => [...baseThemes, ...(customThemes ?? [])],
    [baseThemes, customThemes]
  );

  useEffect(
    function updateThemePreloadData() {
      // Don't update preload data until themes have been loaded and activated
      if (activeThemes == null || customThemes == null) {
        return;
      }

      // Override fill color for certain inline SVGs (the originals are provided
      // by theme-svg.scss)
      overrideSVGFillColors();

      const preloadStyleContent = calculatePreloadStyleContent();

      log.debug2('updateThemePreloadData:', {
        active: activeThemes.map(theme => theme.themeKey),
        custom: customThemes.map(theme => theme.themeKey),
        preloadStyleContent,
        selectedThemeKey,
      });

      setThemePreloadData({
        themeKey: selectedThemeKey,
        preloadStyleContent,
      });
    },
    [activeThemes, selectedThemeKey, customThemes]
  );

  useEffect(() => {
    setValue({
      activeThemes,
      selectedThemeKey,
      themes,
      setSelectedThemeKey,
    });
  }, [activeThemes, selectedThemeKey, themes]);

  return (
    <>
      {activeThemes == null ? null : (
        <>
          {activeThemes.map(theme => (
            <style data-theme-key={theme.themeKey} key={theme.themeKey}>
              {theme.styleContent}
            </style>
          ))}
        </>
      )}
      {value == null ? null : (
        <ThemeContext.Provider value={value}>
          <SpectrumThemeProvider>{children}</SpectrumThemeProvider>
        </ThemeContext.Provider>
      )}
    </>
  );
}

export default ThemeProvider;
