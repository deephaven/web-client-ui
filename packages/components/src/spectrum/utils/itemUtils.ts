import { Key, ReactElement, ReactNode } from 'react';
import type { ItemRenderer } from '@react-types/shared';
import { isElementOfType } from '@deephaven/react-hooks';
import { ensureArray, KeyedItem, SelectionT } from '@deephaven/utils';
import {
  Item,
  ItemElementOrPrimitive,
  ItemProps,
  Section,
  SectionProps,
} from '../shared';
import { PopperOptions } from '../../popper';
import { Text } from '../Text';
import ItemContent from '../ItemContent';

/**
 * `Item.textValue` prop needs to be a non-empty string for accessibility
 * purposes. This is not displayed in the UI.
 */
export const ITEM_EMPTY_STRING_TEXT_VALUE = 'Empty';

/**
 * React Spectrum <Section> supports an `ItemRenderer` function as a child. The
 * DH picker makes use of this internally, but we don't want to support it as
 * an incoming prop.
 */
type SectionPropsNoItemRenderer<T> = Omit<SectionProps<T>, 'children'> & {
  children: Exclude<SectionProps<T>['children'], ItemRenderer<T>>;
};

export type ItemElement<T = unknown> = ReactElement<ItemProps<T>>;
export type SectionElement<T = unknown> = ReactElement<
  SectionPropsNoItemRenderer<T>
>;

export type ItemOrSection<T = unknown> =
  | ItemElementOrPrimitive<T>
  | SectionElement<T>;

// Picker uses `icon` slot. ListView can use `image` or `illustration` slots.
// https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/picker/src/Picker.tsx#L194
// https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/list/src/ListViewItem.tsx#L266-L267
export type ItemIconSlot = 'icon' | 'image' | 'illustration';

/**
 * Augment the Spectrum selection key type to include boolean values.
 * Spectrum collection components already supports this, but the built in types
 * don't reflect it.
 */
export type ItemKey = Key | boolean;

export type ItemSelection = SelectionT<ItemKey>;

/**
 * Augment the Spectrum selection change handler type to include boolean keys.
 * Spectrum components already supports this, but the built in types don't
 * reflect it.
 */
export type ItemSelectionChangeHandler = (key: ItemKey) => void;

export interface MultipleItemSelectionProps {
  selectedKeys?: 'all' | Iterable<ItemKey>;
  defaultSelectedKeys?: 'all' | Iterable<ItemKey>;
  disabledKeys?: Iterable<ItemKey>;

  /**
   * Handler that is called when the selection change.
   * Note that under the hood, this is just an alias for Spectrum's
   * `onSelectionChange`. We are renaming for better consistency with other
   * components.
   */
  onChange?: (keys: ItemSelection) => void;

  /**
   * Handler that is called when the selection changes.
   * @deprecated Use `onChange` instead
   */
  onSelectionChange?: (keys: ItemSelection) => void;
}

export interface NormalizedItemData {
  key?: ItemKey;
  content: ReactNode;
  description?: ReactNode;
  icon?: ReactNode;
  textValue: string | undefined;
}

export interface NormalizedSectionData {
  key?: Key;
  title?: ReactNode;
  items: NormalizedItem[];
}

/**
 * Spectrum collection components support a variety of item types, including
 * strings, numbers, booleans, and more complex React elements. This type
 * represents a normalized form to make rendering items simpler and keep the
 * logic of transformation in separate util methods. It also adheres to the
 * `KeyedItem` interface to be compatible with Windowed data utils
 * (e.g. `useViewportData`).
 */
export type NormalizedItem = KeyedItem<NormalizedItemData, ItemKey>;

export type NormalizedSection = KeyedItem<NormalizedSectionData, Key>;

export type NormalizedItemOrSection<TItemOrSection extends ItemOrSection> =
  TItemOrSection extends SectionElement ? NormalizedSection : NormalizedItem;

export type TooltipOptions = { placement: PopperOptions['placement'] };

/**
 * DH wrappers of Spectrum collection components use a normalized item that
 * includes a `key` prop and an optional `item` prop. This is mostly to support
 * Windowed data where items are created before their data has been loaded (data
 * gets set in the `item` prop). If data has loaded, return its `key`. If not,
 * return the top-level `key` on the normalized item.
 * @param item The normalized item or section
 * @returns The `key` of the item or section
 */
export function getItemKey<
  TItem extends NormalizedItem | NormalizedSection,
  TKey extends TItem extends NormalizedItem
    ? ItemKey
    : TItem extends NormalizedSection
    ? Key
    : undefined,
>(item: TItem | null | undefined): TKey {
  return (item?.item?.key ?? item?.key) as TKey;
}

/**
 * Determine Item `textValue` based on the `textValue` prop or primitive children
 * value.
 * @param item The item to get the text value for
 * @returns The text value of the item
 */
export function getItemTextValue<T>(item: ItemElement<T>): string | undefined {
  if (item.props.textValue == null) {
    const itemKeyStr = item.key == null ? undefined : String(item.key);
    return ['string', 'boolean', 'number'].includes(typeof item.props.children)
      ? String(item.props.children)
      : itemKeyStr;
  }

  return item.props.textValue === ''
    ? ITEM_EMPTY_STRING_TEXT_VALUE
    : item.props.textValue;
}

/**
 * Get the position of the item with the given selected key in a list of items.
 * @param items The items to search
 * @param itemHeight The height of each item
 * @param selectedKey The key of the selected item
 * @param topOffset The top offset of the list
 * @returns The position of the selected item or the top offset if not found
 */
export async function getPositionOfSelectedItemElement<
  TKey extends string | number | boolean | undefined,
>({
  items,
  itemHeight,
  selectedKey,
  topOffset,
}: {
  items: ItemElement[];
  selectedKey: TKey | null | undefined;
  itemHeight: number;
  topOffset: number;
}): Promise<number> {
  let position = topOffset;

  if (selectedKey == null) {
    return position;
  }

  for (let i = 0; i < items.length; i += 1) {
    const item = items[i];
    if (item.key === selectedKey) {
      return position;
    }

    position += itemHeight;
  }

  return topOffset;
}

/**
 * Determine if a node is a Section element.
 * @param node The node to check
 * @returns True if the node is a Section element
 */
export function isSectionElement<T>(
  node: ReactNode
): node is SectionElement<T> {
  return isElementOfType(node, Section);
}

/**
 * Determine if a node is an Item element.
 * @param node The node to check
 * @returns True if the node is an Item element
 */
export function isItemElement<T>(node: ReactNode): node is ItemElement<T> {
  return isElementOfType(node, Item);
}

/**
 * Determine if a node is an Item element containing a child `Text` element with
 * a `slot` prop set to `description`.
 * @param node The node to check
 * @returns True if the node is an Item element with a description
 */
export function isItemElementWithDescription<T>(
  node: ReactNode
): node is ReactElement<ItemProps<T>> {
  if (!isItemElement(node)) {
    return false;
  }

  // If children are wrapped in `ItemContent`, go down 1 level
  const children = isElementOfType(node.props.children, ItemContent)
    ? node.props.children.props.children
    : node.props.children;

  const childrenArray = ensureArray(children);

  const result = childrenArray.some(
    child => isElementOfType(child, Text) && child.props?.slot === 'description'
  );

  return result;
}

/**
 * Determine if a node is an array containing normalized items or sections with
 * keys. Note that this only checks the first node in the array.
 * @param node The node to check
 * @returns True if the node is a normalized item or section with keys array
 */
export function isNormalizedItemsWithKeysList<
  TItemOrSection extends ItemOrSection,
>(
  node:
    | TItemOrSection
    | TItemOrSection[]
    | NormalizedItemOrSection<TItemOrSection>[]
): node is NormalizedItemOrSection<TItemOrSection>[] {
  if (!Array.isArray(node)) {
    return false;
  }

  if (node.length === 0) {
    return true;
  }

  return (
    !isItemOrSection(node[0]) && typeof node[0] === 'object' && 'key' in node[0]
  );
}

/**
 * Determine if an object is a normalized section.
 * @param maybeNormalizedSection The object to check
 * @returns True if the object is a normalized section
 */
export function isNormalizedSection(
  maybeNormalizedSection: NormalizedItem | NormalizedSection
): maybeNormalizedSection is NormalizedSection {
  return (
    maybeNormalizedSection.item != null &&
    'items' in maybeNormalizedSection.item
  );
}

/**
 * Determine if a node is an item or section. Valid types include strings,
 * numbers, booleans, Item elements, and Section elements.
 * @param node The node to check
 * @returns True if the node is an item or section
 */
export function isItemOrSection(node: ReactNode): node is ItemOrSection {
  return (
    typeof node === 'string' ||
    typeof node === 'number' ||
    typeof node === 'boolean' ||
    isItemElement(node) ||
    isSectionElement(node)
  );
}

/**
 * Returns a TooltipOptions object or null if options is false or null.
 * @param options Tooltip options
 * @param placement Default placement for the tooltip if `options` is set
 * explicitly to `true`
 * @returns TooltipOptions or null
 */
export function normalizeTooltipOptions(
  options?: boolean | TooltipOptions | null,
  placement: TooltipOptions['placement'] = 'right'
): TooltipOptions | null {
  if (options == null || options === false) {
    return null;
  }

  if (options === true) {
    return { placement };
  }

  return options;
}

/**
 * Convert a selection of `ItemKey`s to a selection of strings.
 * @param itemKeys The selection of `ItemKey`s
 * @returns The selection of strings
 */
export function itemSelectionToStringSet<
  TKeys extends 'all' | Iterable<ItemKey> | undefined,
  TResult extends TKeys extends 'all'
    ? 'all'
    : TKeys extends Iterable<ItemKey>
    ? Set<string>
    : undefined,
>(itemKeys: TKeys): TResult {
  if (itemKeys == null || itemKeys === 'all') {
    return itemKeys as undefined | 'all' as TResult;
  }

  return new Set([...itemKeys].map(String)) as TResult;
}
