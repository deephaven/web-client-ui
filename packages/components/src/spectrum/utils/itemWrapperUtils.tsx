import { cloneElement, ReactNode } from 'react';
import { Item } from '@adobe/react-spectrum';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { dh as dhIcons } from '@deephaven/icons';
import { isElementOfType } from '@deephaven/react-hooks';
import { ensureArray, NON_BREAKING_SPACE } from '@deephaven/utils';
import {
  isItemElement,
  isSectionElement,
  ItemElement,
  ItemIconSlot,
  ItemOrSection,
  ITEM_EMPTY_STRING_TEXT_VALUE,
  SectionElement,
  TooltipOptions,
} from './itemUtils';
import { ItemContent } from '../ItemContent';
import { Icon } from '../icons';
import { Text } from '../Text';

/**
 * If the given content is a string, wrap it in an Icon component. Otherwise,
 * return the original content. If the key is not found in the dhIcons object,
 * the vsBlank icon will be used.
 * @param maybeIconKey The content to wrap
 * @param slot The slot to use for the Icon component
 * @returns The wrapped content or original content if not a string
 */
export function wrapIcon(
  maybeIconKey: ReactNode,
  slot: ItemIconSlot
): ReactNode {
  // eslint-disable-next-line no-param-reassign
  maybeIconKey = maybeIconKey ?? '';

  if (typeof maybeIconKey !== 'string') {
    return maybeIconKey;
  }

  return (
    <Icon slot={slot}>
      <FontAwesomeIcon icon={dhIcons[maybeIconKey] ?? dhIcons.vsBlank} />
    </Icon>
  );
}

/**
 * Ensure all primitive children are wrapped in `Item` elements and that all
 * `Item` element content is wrapped in `ItemContent` elements to handle text
 * overflow consistently and to support tooltips.
 * @param itemsOrSections The items or sections to wrap
 * @param tooltipOptions The tooltip options to use when wrapping items
 * @returns The wrapped items or sections
 */
export function wrapItemChildren<T>(
  itemsOrSections: ItemOrSection<T> | ItemOrSection<T>[],
  tooltipOptions: TooltipOptions | null
): ItemElement<T> | SectionElement<T> | (ItemElement<T> | SectionElement<T>)[] {
  const itemsOrSectionsArray = ensureArray(itemsOrSections);

  const result = itemsOrSectionsArray.map(item => {
    if (isItemElement(item)) {
      // Item content is already wrapped
      if (isElementOfType(item.props.children, ItemContent)) {
        return item;
      }

      const key = item.key ?? item.props.textValue;
      const textValue =
        item.props.textValue === ''
          ? ITEM_EMPTY_STRING_TEXT_VALUE
          : item.props.textValue;

      // Wrap in `ItemContent` so we can support tooltips and handle text
      // overflow
      return cloneElement(item, {
        ...item.props,
        key,
        textValue,
        children: (
          <ItemContent tooltipOptions={tooltipOptions}>
            {item.props.children}
          </ItemContent>
        ),
      });
    }

    if (isSectionElement(item)) {
      return cloneElement(item, {
        ...item.props,
        key:
          item.key ??
          (typeof item.props.title === 'string' ? item.props.title : undefined),
        children: wrapItemChildren<T>(item.props.children, tooltipOptions),
      });
    }

    const text = String(item);
    const textValue = text === '' ? ITEM_EMPTY_STRING_TEXT_VALUE : text;

    return (
      <Item key={text} textValue={textValue}>
        <ItemContent tooltipOptions={tooltipOptions}>{text}</ItemContent>
      </Item>
    );
  });

  // Keep consistent with original data structure as array or single node
  return Array.isArray(itemsOrSections) ? result : result[0];
}

/**
 * If the given content is a primitive type, wrap it in a Text component.
 * @param content The content to wrap
 * @param slot The slot to use for the Text component
 * @returns The wrapped content or original content if not a primitive type
 */
export function wrapPrimitiveWithText(
  content?: ReactNode,
  slot?: string
): ReactNode {
  // eslint-disable-next-line no-param-reassign
  content = content ?? '';

  if (['string', 'boolean', 'number'].includes(typeof content)) {
    return (
      <Text slot={slot}>
        {/* Non-breaking space is needed to avoid the Text element's height 
        collapsing when content is empty */}
        {content === '' ? NON_BREAKING_SPACE : String(content)}
      </Text>
    );
  }

  return content;
}
