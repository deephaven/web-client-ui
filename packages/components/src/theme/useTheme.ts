import { useContextOrThrow } from '@deephaven/react-hooks';
import { ThemeContext, ThemeContextValue } from './ThemeProvider';

/**
 * Hook to get the current `ThemeContextValue`.
 */
export function useTheme(): ThemeContextValue {
  return useContextOrThrow(
    ThemeContext,
    'No ThemeContext value found. Component must be wrapped in a ThemeContext.Provider'
  );
}

export default useTheme;
