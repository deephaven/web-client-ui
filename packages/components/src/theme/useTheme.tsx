import { useContext } from 'react';
import { ThemeContextType, ThemeContext } from './ThemeProvider';

const useTheme = (): ThemeContextType => useContext(ThemeContext);

export default useTheme;
