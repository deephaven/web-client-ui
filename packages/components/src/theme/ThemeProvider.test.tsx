import React from 'react';
import { act, render } from '@testing-library/react';
import { assertNotNull, TestUtils } from '@deephaven/utils';
import { ThemeContextValue, ThemeProvider } from './ThemeProvider';
import {
  DEFAULT_DARK_THEME_KEY,
  DEFAULT_LIGHT_THEME_KEY,
  ThemeData,
  ThemePreloadData,
} from './ThemeModel';
import {
  calculatePreloadStyleContent,
  getActiveThemes,
  getDefaultBaseThemes,
  getThemePreloadData,
  setThemePreloadData,
} from './ThemeUtils';
import { useTheme } from './useTheme';

const { asMock } = TestUtils;

jest.mock('./ThemeUtils', () => {
  const actual = jest.requireActual('./ThemeUtils');
  return {
    ...actual,
    calculatePreloadStyleContent: jest.fn(),
    getThemePreloadData: jest.fn(actual.getThemePreloadData),
    setThemePreloadData: jest.fn(),
  };
});

const customThemes = [{ themeKey: 'themeA' }] as [ThemeData];
const preloadA: ThemePreloadData = { themeKey: 'themeA' };

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
  expect.hasAssertions();

  asMock(calculatePreloadStyleContent)
    .mockName('calculatePreloadStyleContent')
    .mockReturnValue(':root{mock-preload-content}');

  asMock(getThemePreloadData).mockName('getThemePreloadData');
  asMock(setThemePreloadData).mockName('setThemePreloadData');
});

describe('ThemeProvider', () => {
  const themeContextValueRef = { current: null as ThemeContextValue | null };

  function MockChild() {
    themeContextValueRef.current = useTheme();
    return <div>Child</div>;
  }

  beforeEach(() => {
    themeContextValueRef.current = null;
  });

  it.each([
    [null, null],
    [null, preloadA],
    [customThemes, null],
    [customThemes, preloadA],
  ] as const)(
    'should load themes based on preload data or default: %s, %s',
    (themes, preloadData) => {
      asMock(getThemePreloadData).mockReturnValue(preloadData);

      const component = render(
        <ThemeProvider themes={themes}>
          <MockChild />
        </ThemeProvider>
      );

      assertNotNull(themeContextValueRef.current);

      if (themes == null) {
        expect(themeContextValueRef.current.activeThemes).toBeNull();
      } else {
        expect(themeContextValueRef.current.activeThemes).toEqual(
          getActiveThemes(preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY, {
            base: getDefaultBaseThemes(),
            custom: themes,
          })
        );

        expect(themeContextValueRef.current.selectedThemeKey).toEqual(
          preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY
        );
      }

      expect(component.baseElement).toMatchSnapshot();
    }
  );

  it.each([
    [null, null],
    [null, preloadA],
    [customThemes, null],
    [customThemes, preloadA],
  ] as const)(
    'should set preload data when active themes change: %s, %s',
    (themes, preloadData) => {
      asMock(getThemePreloadData).mockReturnValue(preloadData);

      render(
        <ThemeProvider themes={themes}>
          <MockChild />
        </ThemeProvider>
      );

      if (themes == null) {
        expect(setThemePreloadData).not.toHaveBeenCalled();
      } else {
        expect(setThemePreloadData).toHaveBeenCalledWith({
          themeKey: preloadData?.themeKey ?? DEFAULT_DARK_THEME_KEY,
          preloadStyleContent: calculatePreloadStyleContent(),
        });
      }
    }
  );

  describe.each([null, customThemes])('setSelectedThemeKey: %s', themes => {
    it.each([DEFAULT_LIGHT_THEME_KEY, customThemes[0].themeKey])(
      'should change selected theme: %s',
      themeKey => {
        const component = render(
          <ThemeProvider themes={themes}>
            <MockChild />
          </ThemeProvider>
        );

        assertNotNull(themeContextValueRef.current);

        if (themes == null) {
          expect(themeContextValueRef.current.activeThemes).toBeNull();
        } else {
          act(() => {
            themeContextValueRef.current!.setSelectedThemeKey(themeKey);
          });

          expect(themeContextValueRef.current.activeThemes).toEqual(
            getActiveThemes(themeKey, {
              base: getDefaultBaseThemes(),
              custom: themes,
            })
          );

          expect(themeContextValueRef.current.selectedThemeKey).toEqual(
            themeKey
          );
        }

        expect(component.baseElement).toMatchSnapshot();
      }
    );
  });
});
