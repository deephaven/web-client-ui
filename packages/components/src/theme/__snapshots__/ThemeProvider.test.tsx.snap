// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider setSelectedThemeKey: [ [Object] ] should change selected theme: default-light 1`] = `
<body>
  <div>
    <style
      data-theme-key="default-light"
    >
      ./theme-light-palette.css?raw
./theme-light-semantic.css?raw
./theme-light-semantic-chart.css?raw
./theme-light-semantic-editor.css?raw
./theme-light-semantic-grid.css?raw
./theme-light-components.css?raw
    </style>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider setSelectedThemeKey: [ [Object] ] should change selected theme: themeA 1`] = `
<body>
  <div>
    <style
      data-theme-key="default-dark"
    >
      ./theme-dark-palette.css?raw
./theme-dark-semantic.css?raw
./theme-dark-semantic-chart.css?raw
./theme-dark-semantic-editor.css?raw
./theme-dark-semantic-grid.css?raw
./theme-dark-components.css?raw
    </style>
    <style
      data-theme-key="themeA"
    />
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider setSelectedThemeKey: null should change selected theme: default-light 1`] = `
<body>
  <div>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider setSelectedThemeKey: null should change selected theme: themeA 1`] = `
<body>
  <div>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider should load themes based on preload data or default: [ [Object] ], { themeKey: 'themeA' } 1`] = `
<body>
  <div>
    <style
      data-theme-key="default-dark"
    >
      ./theme-dark-palette.css?raw
./theme-dark-semantic.css?raw
./theme-dark-semantic-chart.css?raw
./theme-dark-semantic-editor.css?raw
./theme-dark-semantic-grid.css?raw
./theme-dark-components.css?raw
    </style>
    <style
      data-theme-key="themeA"
    />
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider should load themes based on preload data or default: [ [Object] ], null 1`] = `
<body>
  <div>
    <style
      data-theme-key="default-dark"
    >
      ./theme-dark-palette.css?raw
./theme-dark-semantic.css?raw
./theme-dark-semantic-chart.css?raw
./theme-dark-semantic-editor.css?raw
./theme-dark-semantic-grid.css?raw
./theme-dark-components.css?raw
    </style>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider should load themes based on preload data or default: null, { themeKey: 'themeA' } 1`] = `
<body>
  <div>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`ThemeProvider should load themes based on preload data or default: null, null 1`] = `
<body>
  <div>
    <div
      class="spectrum-theme-provider spectrum_b37d53 spectrum_2a241c spectrum--light_a40724 theme-spectrum-palette theme-spectrum-alias spectrum--large_c40598 spectrum_9e130c spectrum--medium_9e130c spectrum--large_9e130c spectrum--darkest_9e130c spectrum--dark_9e130c spectrum--light_9e130c spectrum--lightest_9e130c"
      dir="ltr"
      lang="en-US"
      style="isolation: isolate; color-scheme: light;"
    >
      <div>
        Child
      </div>
    </div>
  </div>
</body>
`;
